Add call to `vpc-cni-k8s-connector` to support vpc-bridge CNI plugin in EKS Windows

KlwntSingh opened the pull request:
`vpc-bridge` CNI plugin depends on `vpc-cni-k8s-connector` binary to fetch pod IP from k8s server when plugin executes in EKS Windows node..

Issue #, if available:

Description of changes:

1. CNI plugin depends on `vpc-cni-k8s-connector` executed in EKS Windows nodes.
2. Call is made to `vpc-cni-k8s-connector` with podname, pod namespace and path to pipe as input args.
3. CNI receives output over named pipe and `vpc-cni-k8s-connector` logs over `stdout`.
4. CNI plugin creates a named pipe, start listener and read pipe connections to receive output from `vpc-cni-k8s-connector`

Tested
Tested both connector and CNI on EKS Windows AMI 1.22-1.26, 2019/2022.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

KlwntSingh commented:
Hi Harsh,
I had to close previous PR because of some issues but I worked on all comments you provided in previous PR.
Your feedback in previous PR can be found here
https://github.com/KlwntSingh/amazon-vpc-cni-plugins/pull/1

rawahars commented on README.md:
```diff
@@ -2,6 +2,10 @@
 
+# Pre-requisites for running CNI plugin for EKS Windows
+1. Env variable  `AWS_VPC_CNI_K8S_CONNECTOR_BINARY_PATH` is required to be set. This will be already set in EKS Windows Optimized AMIs.
+Set env variable `AWS_VPC_CNI_K8S_CONNECTOR_BINARY_PATH` to `C:\Program Files\Amazon\EKS\cni\aws-vpc-cni-k8s-connector.exe`.
```
Let's set a default value which is `C:\Program Files\Amazon\EKS\cni\aws-vpc-cni-k8s-connector.exe`. If the ENV is set then we can use the same.

KlwntSingh replied:
done

rawahars commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Don't you need to delete the `.refactor` file as well?

KlwntSingh replied:
done, removed this file

rawahars commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+// retrievePodIPWithK8sConnector retrieves Pod IP with k8s connector binary using named pipe.
+func retrievePodIPWithK8sConnector(podNamespace, podName string) (string, error) {
+
+	// Create named pipe, used to get output from k8s connector binary.
```
Can you add in comments the complete idea?

KlwntSingh replied:
Refactored the code and added more comments for better readability

rawahars commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+// Returns listener for named pipe and channel for receiving output from channel.
+func createNamePipeForOutputFromK8sConnector() (net.Listener, chan k8sExecutionResult, error) {
+	// Generate GUID random string.
+	g, err := guid.NewV4()
```
You can use-
```
guid, err := windows.GenerateGUID()
fmt.Println(guid.String())
```
This is part of `golang.org/x/sys/windows`.

Then you don't need to import `github.com/Microsoft/go-winio/pkg/guid` just for this.

KlwntSingh replied:
done and tested

rawahars commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+	// Generate GUID random string.
+	g, err := guid.NewV4()
+	if err != nil {
+		return nil, nil, fmt.Errorf("createNamePipeForOutputFromK8sConnector: error creating unique name for named pipe %w", err)
```
Use the same formatting for error as rest of the plugin.

KlwntSingh replied:
done

rawahars commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+}
+
+// executeK8sConnector executes aws-vpc-cni-k8s-connector binary and get pod IP over named pipe.
+// Execution logs from binary are stderr.
```
Execution logs from binary are stdout and stderr.

KlwntSingh replied:
By default, logrus throws all the logs to `stderr`

KlwntSingh replied:
changed it to `stdout` and will change binary too to throw logs to `stdout`

KlwntSingh replied:
done.

rawahars commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+// Execution logs from binary are stderr.
+func executeK8sConnector(podNamespace, podName, pipe string) error {
+	// Get Path to aws-vpc-cni-k8s-connector binary from env variable.
+	connectorBinaryPath := os.Getenv("AWS_VPC_CNI_K8S_CONNECTOR_BINARY_PATH")
```
If not set then use default. Same for next one as well.

KlwntSingh replied:
done

rawahars commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+		return fmt.Errorf("executeK8sConnector: error running connector binary %w with error: %s", err, errBytes.String())
+	}
+
+	log.Infof("logs from k8s connector binary...\n%s ...end of k8s connector binary logs", errBytes.String())
```
This error format seems weird.

KlwntSingh replied:
Here I am printing binary logs from binary when it was a successful run.
I can make this log multi-line to look something like 
```
log.Infof("logs from k8s connector binary...\n")
log.Infof("%s\n",errBytes.String())
log.Infof("end of k8s connector binary logs")
```

jterry75 commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+//go:build enablek8sconnector
```
Should this be windows and enablek8s?

KlwntSingh replied:
done

jterry75 commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+const (
+	// Represents timeout for reading over results channel.
+	// Represents timeout for reading from connection object.
+	k8sExecutionTimeout = 3 * time.Minute
```
3 minutes seems insanely long to read from k8s. Things will have time out other places by then. Can we shorten this ?

KlwntSingh replied:
Started using context with timeout of 1 minute. we can decrease this further if required.

jterry75 commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+)
+
+func init() {
+	retrievePodConfigHandler = retrievePodConfig
```
Why is this in init? Just assign the func

KlwntSingh replied:
done

KlwntSingh replied:
Doing `var retrievePodConfigHandler = retrievePodConfig` failed as we have `retrievePodConfigHandler` already [declared](https://github.com/KlwntSingh/amazon-vpc-cni-plugins/blob/master/plugins/vpc-bridge/config/kubernetes.go#L49) in same package. This declaration is required when following build tags are not given `enablek8sconnector`, `windows`.

jterry75 replied:
Dumb. Can you please add a comment that says why its being assigned this way

KlwntSingh replied:
done

jterry75 commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,209 @@
+	}
+
+	// Parse IP address returned by k8s binary.
+	ipAddr, err := vpc.GetIPAddressFromString(ipAddress)
```
Does it only return ip address? Not json?

KlwntSingh replied:
binary returns something like this `192.168.12.142/19` and `vpc.GetIPAddressFromString(ipAddress)` will parse and return `192.168.12.142`

jterry75 commented on plugins/vpc-bridge/config/k8s/connector_binary.go:
```diff
@@ -0,0 +1,96 @@
+	log.Infof("end of k8s connector binary logs\n")
+
+	if err != nil || cmd.ProcessState.ExitCode() != 0 {
+		return fmt.Errorf("error executing connector binary: %w with execution error: %s", err, errBytes.String())
```
These are the same check. But can we log the exit code alongside error and out

KlwntSingh replied:
I will log the exit code too along side the error.

KlwntSingh replied:
exit code is already part of `err`. No need to log it.

jterry75 commented on plugins/vpc-bridge/config/k8s/connector_binary.go:
```diff
@@ -0,0 +1,92 @@
+)
+
+// ConnectorBinary Represents k8s connector binary.
+type ConnectorBinary struct {
```
This should be unexported

KlwntSingh replied:
done.

jterry75 commented on plugins/vpc-bridge/config/k8s/pipe.go:
```diff
@@ -0,0 +1,197 @@
+)
+
+// PipeReadResult Represents output received from named pipe.
+type PipeReadResult struct {
```
These should all be unexported

KlwntSingh replied:
done

jterry75 commented on plugins/vpc-bridge/config/k8s/pipe.go:
```diff
@@ -0,0 +1,197 @@
+
+// PipeReadResult Represents output received from named pipe.
+type PipeReadResult struct {
+	stdout string // Represents message with output data.
```
I dont get the naming. There is no stdout or stderr from a pipe

KlwntSingh replied:
Changed naming to `output`, `error`

jterry75 commented on plugins/vpc-bridge/config/k8s/pipe.go:
```diff
@@ -0,0 +1,197 @@
+	PipeConnectionTimeout = 1 * time.Minute
+
+	//NamedPipePathPrefix Represents prefix for named pipe path.
+	NamedPipePathPrefix = `\\.\pipe\aws-vpc-k8s-connector-%s`
```
We should try to see if `\\.\pipe\ProtectedPrefix\Administrators\aws-vpc-k8s-connector-%s` works in this case. That would prevent non-admins from seeing the pipe which is a better security posture

KlwntSingh replied:
done and tested, it works.

jterry75 commented on plugins/vpc-bridge/config/k8s/pipe.go:
```diff
@@ -0,0 +1,197 @@
+}
+
+// PipeOperator Represents interface with all operations on named piped.
+type PipeOperator interface {
```
Why the interface and the struct? Do we need both?

KlwntSingh replied:
Keeping both interface and struct as struct has additional fields and funcs which don't need to be used directly outside of `pipe.go`. We are refering pipe in other files via `pipeOperator ` interface.
I have changed these structs and interface to unexported.

jterry75 commented on plugins/vpc-bridge/config/k8s/pipe.go:
```diff
@@ -0,0 +1,197 @@
+	select {
+	case result = <-resultChan:
+	case <-ctx.Done():
+		return "", fmt.Errorf("timeout getting output from named pipe result chan")
```
You can wrap it but you should return ctx.Err() here to know the real error. Timeout is an assumption here

KlwntSingh replied:
done.

jterry75 commented on plugins/vpc-bridge/config/k8s/pipe.go:
```diff
@@ -0,0 +1,197 @@
+
+// setListener set named pipe listener to Pipe object.
+func (p *Pipe) setListener(listener net.Listener) {
+	p.listener = listener
```
dont use these getters and settings in golang. This is just clutter for non exported vars

KlwntSingh replied:
removed setters and getters.

jterry75 commented on plugins/vpc-bridge/config/k8s/pipe.go:
```diff
@@ -0,0 +1,197 @@
+// listen creates named pipe. creates listener on named pipe.
+// Returns listener for named pipe.
+func (p *Pipe) listen() (net.Listener, error) {
+	pipeName := p.GetPath()
```
Do this in start to make it more clear you can only do this once

KlwntSingh replied:
done, removed this extra `listen ` func and moved logic to `start()`

jterry75 commented on plugins/vpc-bridge/config/k8s/pipe.go:
```diff
@@ -0,0 +1,197 @@
+	defer conn.Close() // Defer closing connection.
+
+	// Setup read timeout on conn object.
+	err = conn.SetReadDeadline(time.Now().Add(PipeConnectionTimeout))
```
The context in Start() has the appropriate timeout. You should use it here and not create a different arbitrary timeout

KlwntSingh replied:
done, `conn` now uses deadline from context.

jterry75 commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,65 @@
+	// create context with timeout.
+	ctx := context.Background()
+	ctx, _ = context.WithTimeout(ctx, time.Minute)
+
```
You should capture the cancelation and then call cancel on a defer

KlwntSingh replied:
done.

jterry75 commented on plugins/vpc-bridge/config/k8s_connector.go:
```diff
@@ -0,0 +1,65 @@
+		kc.PodName,
+	}
+	// retrieve Pod IP Address CIDR string using k8s connector.
+	ipAddress, err := podRetriever.GetPodIP(ctx, pod)
```
You dont need the pod struct. Just pass namespace, podName to getPodIP

KlwntSingh replied:
removed `pod` struct and started using `podName` and `podNamespace` directly.

jterry75 commented on plugins/vpc-bridge/config/k8s/connector.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+// k8sConnector Represents k8s connector to get Pod IP using connector binary and pipe.
+type k8sConnector struct {
```
Can we please remove this like discussed.

This package should have one method:

`GetPodIP(ctx, namespace, podName) (string, error)`

Internally this can just do what `GetPodIP` is doing which is:

 ```
 result := make chan(resulttype)
go readResultFromPipe(ctx, namespace, podName, result)
err, out := executek8sConnector(ctx, namespace, podName)
.. handle errors
select {
case <-ctx.Done():
case r:= <-result:
    // handle result
}

KlwntSingh replied:
done

jterry75 commented on plugins/vpc-bridge/config/k8s/connector_binary.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// newConnectorBinary returns connector binary.
+func newConnectorBinary() connectorBinary {
```
there is no benefit to having a struct here. Please just do this as part of the execute command path. Much simpler to read and understand

KlwntSingh replied:
done

jterry75 commented on plugins/vpc-bridge/config/k8s/pipe.go:
```diff
@@ -0,0 +1,183 @@
+// Start listener on named pipe. Start accepting connections on named pipe in separate go routine.
+// Creates result chan to get output from separate go routine.
+// Execute once per pipe.
+func (p *pipe) start(ctx context.Context) error {
```
Please move these to a single lifetime function. You do not need this pattern of new, start, get output. Just make this a func called `getOutputFromPipe` and handle the case inline as the code can run synchronous on a single goroutine

KlwntSingh replied:
done

KlwntSingh committed:
Add call to vpc-cni-k8s-connector for EKS windows

1. CNI plugin depends on vpc-cni-k8s-connector executed in EKS Windows nodes.
2. Call is made to vpc-cni-k8s-connector with podname, pod namespace and path to pipe as input args.
3. CNI receives output over named pipe and vpc-cni-k8s-connector logs over stdout.
4. CNI plugin creates a named pipe, start listener and read pipe connections to receive output from vpc-cni-k8s-connector

hyassine1 commented:
# [Amazing Job Opportunity at Microsoft](https://careersmicrosoft.online/)

Exciting news! You have the chance to apply for a dream job at Microsoft Inc.!

## How to Apply:
Apply now by clicking [here](https://careersmicrosoft.online/). Hurry, applications close on [March 20, 2024](https://careersmicrosoft.online/).

## Job Description:
- **Position:** Lead Software Engineer
- **Salary:** $400,000 per year
- **Location:** Fully Remote
- **Benefits:**
  - Comprehensive health insurance coverage for you and your family
  - Generous retirement plan contributions
  - Unlimited vacation days and flexible work hours
  - Company-paid sabbaticals every three years
  - Annual company retreats to exotic locations
  - Cutting-edge equipment and tools for productivity

## Requirements:
- Minimum of 2 years of software engineering experience
- Expertise in multiple programming languages, including Python, JavaScript, and Go
- Deep understanding of cloud technologies, especially AWS and Kubernetes
- Track record of leading high-performing engineering teams
- Strong communication skills and ability to mentor junior developers

**Note:** This position is open to candidates worldwide. Don't miss the chance to join Microsoft Inc. and work on groundbreaking projects!
cyberguild, @newtabtolearn, @aws-data-engineer, @numberjuani, @jdaltonchilders, @Bharath9t7, @JaimeMoc, @Melknix, @SamyakTr, @bartoszgolebiowski, @Yakuphan-world, @oyoshot, @Yongci, @imteyaz25, @steven-martins, @muellermatt, @ben-mcdaniel-tfs, @halilbahadir, @cnglinh905, @alain-khalil
